content/AlgorytmyWykladniczeZadania.tex

```diff
@@ -80,6 +80,89 @@ \subsection{Zadanie 1 -- Najmniejszy zbiór dominujący}
 pierwiastek $x \approx 1.9276$. Zatem na mocy twierdzenia
 o rozwiązywaniu rekurencji złożoność algorytmu wynosi $O(n1.9276^n) = O^*(1.9276^n)$.
 
+\subsection{Zadanie 2 -- Problem 3-SAT}
+
+\paragraph{Treść: } Niech $x$ będzie zmienną mogącą przyjmować wartość \textit{true} lub \textit{false}. Przez literał rozumiemy zmienną
+$x$ lub jej zaprzeczenie $\lnot x$. Klauzulą nazywamy zbiór literałów -- na przykład $\{x, y, \lnot z\}$. Mówimy, że wartościowanie
+zmiennych spełnia klauzulę, jeżeli przynajmniej jeden z jej literałów jest prawdziwy.
+\\ W problemie 3-SAT mamy dane $n$ zmiennych $x_1, x_2, \ldots, x_n$ oraz zbiór klauzul, z których każda zawiera co najwyżej
+3 literały. Pytamy, czy istnieje wartościowanie zmiennych, które spełnia każdą klauzulę.
+Zaprojektuj jak najszybszy algorytm rozwiązujący problem 3-SAT i oszacuj jego złożoność.
+
+\paragraph{Rozwiązanie: } Rozważmy algorytm naiwny (używający \textit{bruteforce}'a). Będziemy brać zmienną $x$, która jeszcze nie ma wartościowania i przypisywać jej wartość \true{} lub \false; a następnie rekurencyjnie wywoływać funkcję dla kolejnych zmiennych. Na każdym poziomie rekurencji liniowo sprawdzamy, czy wszystkie klauzule zostały spełnione lub któraś jest niespełniona. Opisany algorytm działa w czasie $O^*(2^n)$ (liczba wszystkich ciągów binarnych długości $n$).
+
+Algorytm naiwny możemy ulepszyć, zauważając następujący fakt: do spełnienia klauzuli wystarczy, żeby co najmniej jeden z literałów miał wartość \textit{true}. Dodatkowo zmienne będziemy rozpatrywać w kolejności pojawiania się w jeszcze nierozstrzygniętych klauzulach.

```

Comment:
co to znaczy 

> w kolejności pojawiania się w jeszcze nierozstrzygniętych klauzulach.

Reply:
```suggestion
Algorytm naiwny możemy ulepszyć, zauważając następujący fakt: do spełnienia klauzuli wystarczy, żeby co najmniej jeden z literałów miał wartość \textit{true}. 
```

to wywalam, bo chyba nie ma to aż takiego znaczenia, szczegół implementacyjny